### PR TITLE
Remove signal handler registration from pkg/kubelet

### PR DESCRIPTION
**Pull request by frobware**
**What this PR does / why we need it**:

The goal of this change is to remove the registration of signal
handling from pkg/kubelet (i.e., library code). 
 
If you register a signal handler in `main()` to aid in a controlled
and deliberate exit then the handler registered in `pkg/kubelet` often
wins and the process exits immediately. This means all other signal
handler registrations are currently racy if `DockerServer.Start()` is
directly or indirectly invoked.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

This change also removes another signal handler registration from
`NewAPIServerCommand()`; a stop channel is now passed to this
function.

**Release note**:

### CODE REVIEW

File: cmd/hyperkube/main.go [hunk 1]
```diff
@@ -82,12 +83,12 @@ func commandFor(basename string, defaultCommand *cobra.Command, commands []func(
 	return defaultCommand
 }
 
-// NewCmdRequestProject implement the OpenShift cli RequestProject command.
-func NewHyperKubeCommand() (*cobra.Command, []func() *cobra.Command) {
+// NewHyperKubeCommand implement the OpenShift cli RequestProject command.
```
Comment (derekwaynecarr) [inline cmd/hyperkube/main.go]: this comment is not right...

Comment (frobware) [reply cmd/hyperkube/main.go]: > this comment is not right...

Done.

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -106,7 +105,7 @@ const etcdRetryLimit = 60
 const etcdRetryInterval = 1 * time.Second
 
 // NewAPIServerCommand creates a *cobra.Command object with default parameters
-func NewAPIServerCommand() *cobra.Command {
+func NewAPIServerCommand(stopCh <-chan struct{}) *cobra.Command {
```
Comment (derekwaynecarr) [inline cmd/kube-apiserver/app/server.go]: it is not obvious to me why we are changing this in this pr.

it seems unrelated to the kubelet changes needing to pass a channel down to the dockershim.

**Review by derekwaynecarr — CHANGES_REQUESTED** [review]
can you explain why we need to make changes to API server?

File: cmd/kube-apiserver/app/server.go (hunk 2 above)
Comment (frobware) [reply cmd/kube-apiserver/app/server.go]: In `cmd/hyperkube/main` we setup the signal handler using `server.SetupSignalHandler()`. This function can only be called one; subsequent calls deliberately fail:

    func SetupSignalHandler() (stopCh <-chan struct{}) {
            close(onlyOneSignalHandler) // panics when called twice
            ...

`NewHyperKubeCommand()` directly calls `NewAPIServerCommand()` and prior to this PR `NewAPIServerCommand()` also called `SetupSignalHandler()`. Without removing the secondary registration in `NewAPIServerCommand()` the registration would happen twice, resulting in a `panic on closed channel`.

To fix this we now pass the stopCh in from the roots of the call trees rather than letting it happen part-way down in `NewAPIServerCommand()`. 

And this is why this PR also changed the following files:

```console
$ ag NewAPIServerCommand

File: cmd/genman/gen_kube_man.go
66:27:		apiserver := apiservapp.NewAPIServerCommand(server.SetupSignalHandler())

File: cmd/genkubedocs/gen_kube_docs.go
57:27:		apiserver := apiservapp.NewAPIServerCommand(server.SetupSignalHandler())

File: cmd/kube-apiserver/app/server.go
107:4:// NewAPIServerCommand creates a *cobra.Command object with default parameters
108:6:func NewAPIServerCommand(stopCh <-chan struct{}) *cobra.Command {

File: cmd/kube-apiserver/apiserver.go
41:17:	command := app.NewAPIServerCommand(server.SetupSignalHandler())
```

**Comment by derekwaynecarr** [comment]
@frobware thank you for the explanation.

can you squash commits then this is lgtm.

**Comment by frobware** [comment]
> can you squash commits then this is lgtm.

done.

**Review by derekwaynecarr — APPROVED** [review]

File: cmd/kubelet/app/server.go [hunk 3]
```diff
@@ -722,7 +723,13 @@ func run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) (err error) {
 	// If systemd is used, notify it that we have started
 	go daemon.SdNotify(false, "READY=1")
 
-	<-done
+	select {
```
Comment (liggitt) [inline cmd/kubelet/app/server.go]: so `done`  is a lockfile-based method for telling the kublet to shut down, and stopCh is a signal-based mechanism? I'm surprised we weren't already passing `done` down into RunKubelet to get graceful shutdowns

Comment (liggitt) [reply cmd/kubelet/app/server.go]: should we be calling `SdNotify(false, "STOPPING=1")` as part of shutdown?

do we have to wait until RunKubelet has finished responding to the closed stopCh before we return?

File: pkg/kubelet/kubelet.go [hunk 4]
```diff
@@ -617,9 +618,8 @@ func NewMainKubelet(kubeCfg *kubeletconfiginternal.KubeletConfiguration,
 		glog.V(5).Infof("RemoteRuntimeEndpoint: %q, RemoteImageEndpoint: %q",
 			remoteRuntimeEndpoint,
 			remoteImageEndpoint)
-		glog.V(2).Infof("Starting the GRPC server for the docker CRI shim.")
```
[the remark this answers is not in the document]
Comment (frobware) [reply pkg/kubelet/kubelet.go]: > why the lost log line?

Restored.

**Comment by liggitt** [comment]
had a couple questions around the kubelet shutdown bits, but this looks like an improvement to me as-is

File: cmd/kubelet/app/server.go (hunk 3 above)
Comment (frobware) [reply cmd/kubelet/app/server.go]: > should we be calling SdNotify(false, "STOPPING=1") as part of shutdown?

That I don't know; I'm speculating that we don't see this in the tree at all because we didn't really gracefully shutdown before.

**Comment by frobware** [comment]
@derekwaynecarr please could you LGTM again; addressed review issues and had to rebase.

File: cmd/kubelet/app/server.go (hunk 3 above)
Comment (smarterclayton) [reply cmd/kubelet/app/server.go]: You don't have to call the shutdown signal.
